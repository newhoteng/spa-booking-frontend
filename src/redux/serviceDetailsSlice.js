import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const getUrl = 'http://127.0.0.1:3001/api/v1/spa_services';

const initialState = {
  serviceDetails: [], // Correctly initialized as an array
  isLoading: true,
  isError: undefined,
};

export const fetchServiceDetails = createAsyncThunk(
  'serviceDetails/fetchServiceDetails',
  async (serviceId) => { // Pass the serviceId parameter
    try {
      const response = await axios.get(`${getUrl.replace(':id', serviceId)}`);
      return response.data;
    } catch (error) {
      throw new Error(error.message);
    }
  },
);

export const serviceDetailsSlice = createSlice({
  name: 'serviceDetails',
  initialState,
  reducers: {

  },
  extraReducers(builder) {
    builder
      .addCase(fetchServiceDetails.pending, (state) => {
        state.isLoading = true;
        state.isError = false;
      })
      .addCase(fetchServiceDetails.fulfilled, (state, action) => {
        state.isLoading = false;
        state.serviceDetails = action.payload;
      })
      .addCase(fetchServiceDetails.rejected, (state) => {
        state.isLoading = false;
        state.isError = true;
      });
  },
});

export const selectServiceDetails = (state) => state.serviceDetails.serviceDetails;

export default serviceDetailsSlice.reducer;
