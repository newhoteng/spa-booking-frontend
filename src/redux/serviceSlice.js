import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import URL from '../consts';

const initialState = {
  services: [],
  isLoading: true,
  isError: false,
};

// Fetch all services
export const fetchAllServices = createAsyncThunk(
  'services/fetchAllServices',
  async () => {
    const response = await axios.get(URL);
    return response.data;
  },
);

// Add a new service
export const addService = createAsyncThunk(
  'services/addService',
  async (serviceData) => {
    const response = await axios.post(URL, serviceData);
    return response.data;
  },
);

// Remove a service
export const removeService = createAsyncThunk('services/RemoveService', async (id, thunkAPI) => {
  try {
    await axios.delete(`http://localhost:3001/api/v1/spa_services/${id}`);
    return {};
  } catch (error) {
    return thunkAPI.rejectWithValue('something went wrong');
  }
});

export const serviceSlice = createSlice({
  name: 'services',
  initialState,
  reducers: {
    toggleService: (state, action) => {
      const itemId = action.payload;
      state.services = state.services.map((service) => {
        if (service.id !== itemId) return service;
        return { ...service, is_removed: !service.is_removed };
      });
    },
  },
  extraReducers(builder) {
    builder
      .addCase(fetchAllServices.pending, (state) => {
        state.isLoading = true;
        state.isError = false;
      })
      .addCase(fetchAllServices.fulfilled, (state, action) => {
        state.isLoading = false;
        state.isError = false;
        state.services = action.payload;
      })
      .addCase(fetchAllServices.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = action.payload;
      })
      .addCase(addService.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(addService.fulfilled, (state, action) => {
        state.isLoading = false;
        // state.isError = false;
        // Add the newly created service to the services array
        state.services.push(action.payload);
      })
      .addCase(addService.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = action.payload;
      });
  },
});

export const { toggleService } = serviceSlice.actions;
export default serviceSlice.reducer;
