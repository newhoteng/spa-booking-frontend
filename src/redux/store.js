import { configureStore } from '@reduxjs/toolkit';
import reservationsReducer from './reservations/reservationsSlice';
import servicesReducer from './serviceSlice';
import serviceDetailsReducer from './serviceDetailsSlice';

const store = configureStore({
  reducer: {
    userReservations: reservationsReducer,
    services: servicesReducer,
    serviceDetails: serviceDetailsReducer,
  },
});

export default store;
