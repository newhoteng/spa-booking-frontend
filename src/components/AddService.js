import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { addService } from '../redux/serviceSlice';
import styles from '../styles/AddService.module.css';

const AddService = () => {
  // Initialize Redux dispatch
  const dispatch = useDispatch();

  // Define state variables for form inputs and submission status
  const [name, setName] = useState('');
  const [image, setImage] = useState('');
  const [description, setDescription] = useState('');
  const [duration, setDuration] = useState('');
  const [price, setPrice] = useState('');
  const [isSubmitted, setIsSubmitted] = useState(false);

  // Handle form submission
  const handleSubmit = async (e) => {
    e.preventDefault();
    // Dispatch the 'addService' action with service details
    const response = await dispatch(
      addService({
        name,
        image,
        description,
        duration,
        price,
      }),
    );
    // If the action was successful, set the submission status to true
    if (response.type === 'services/addService/fulfilled') {
      setIsSubmitted(true);
    }
     // Clear form fields after submission
    setName('');
    setImage('');
    setDescription('');
    setDuration('');
    setPrice('');
  };

  const handleContinueAdding = () => {
    setIsSubmitted(false);
  };

  return (
    <div className={styles['add-service-container']}>
      {isSubmitted ? (
        <div className={styles['success-message']}>
          <p>Your service has been added successfully!</p>
          <div className={styles['success-actions']}>
            <button type="button" onClick={handleContinueAdding}>
              Continue Adding
            </button>
            <Link to="/">
              <button type="button">Go to Home</button>
            </Link>
          </div>
        </div>
      ) : (
        <>
          <h2 className={styles.header}>ADD SERVICE</h2>
          <form onSubmit={handleSubmit} className={styles['add-service-form']}>
            <div className={styles['add-service-form-group']}>
              <input
                type="text"
                className={styles.placeholder}
                id="name"
                value={name}
                onChange={(e) => setName(e.target.value)}
                placeholder="Enter Service Name"
                required
              />
              <input
                type="text"
                className={styles.placeholder}
                id="image"
                value={image}
                onChange={(e) => setImage(e.target.value)}
                placeholder="Add Image URL"
              />
              <input
                type="textarea"
                id="description"
                className={`${styles.placeholder_desc} ${styles.placeholder}`}
                value={description}
                onChange={(e) => setDescription(e.target.value)}
                placeholder="Enter Service Description"
                required
              />
              <input
                type="number"
                className={styles.placeholder}
                id="duration"
                value={duration}
                onChange={(e) => setDuration(e.target.value)}
                placeholder="Enter Service Duration in Minutes"
                required
              />
              <input
                type="number"
                className={styles.placeholder}
                id="price"
                value={price}
                onChange={(e) => setPrice(e.target.value)}
                placeholder="Enter Service Price"
                required
              />
            </div>
            <button type="submit" className={styles.add_services}>Add Service</button>
          </form>
        </>
      )}
    </div>
  );
};

export default AddService;
