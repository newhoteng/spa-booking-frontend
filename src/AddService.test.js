import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux'; // Import the Provider
import { createStore } from 'redux'; // Import createStore from Redux
import AddService from './components/AddService';
import rootReducer from './redux/serviceSlice'; // Import your rootReducer

test('AddService component matches snapshot when not submitted', () => {
  // Create a mock Redux store
  const store = createStore(rootReducer);

  const { container } = render(
    <Provider store={store}>
      {/* Wrap component with Provider */}
      <AddService />
    </Provider>,
  );

  // Use snapshot testing to compare the rendered component with a stored snapshot.
  expect(container).toMatchSnapshot();
});

test('AddService component matches snapshot when submitted', () => {
  const store = createStore(rootReducer);

  render(
    <Provider store={store}>
      <AddService />
    </Provider>,
  );
});
