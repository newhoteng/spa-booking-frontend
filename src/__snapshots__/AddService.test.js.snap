// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddService component matches snapshot when not submitted 1`] = `
<div>
  <div
    class="add-service-container"
  >
    <h2
      class="header"
    >
      ADD SERVICE
    </h2>
    <form
      class="add-service-form"
    >
      <div
        class="add-service-form-group"
      >
        <input
          class="placeholder"
          id="name"
          placeholder="Enter Service Name"
          required=""
          type="text"
          value=""
        />
        <input
          class="placeholder"
          id="image"
          placeholder="Add Image URL"
          type="text"
          value=""
        />
        <input
          class="placeholder_desc placeholder"
          id="description"
          placeholder="Enter Service Description"
          required=""
          type="textarea"
          value=""
        />
        <input
          class="placeholder"
          id="duration"
          placeholder="Enter Service Duration in Minutes"
          required=""
          type="number"
          value=""
        />
        <input
          class="placeholder"
          id="price"
          placeholder="Enter Service Price"
          required=""
          type="number"
          value=""
        />
      </div>
      <button
        class="add_services"
        type="submit"
      >
        Add Service
      </button>
    </form>
  </div>
</div>
`;
